Grammar fixes

stefkauf opened the pull request.

stefkauf committed:
Re-implemented conversion of CFG to Chomsky Normal Form

Conversion to Chomsky Normal Form (CNF), implemented as the
'.chomsky_normal_form()' method in grammar.CFG instances, was
incomplete. In particular, it could not deal with empty productions
or mixed productions (with terminals and non-terminals co-occuring
on the right-hand side). The new implementation removes those
shortcomings. Like in the earlier version, the individual steps of
conversion are carried out by calling class methods.

stefkauf committed:
Added Stefan Kaufmann to AUTHORS.md

tomaarsen commented:
Hey @stefkauf! You might've noticed that the tests are failing for this. The reason for this is explained in this comment: https://github.com/nltk/nltk/pull/2822#issuecomment-928037649. I hope this helps.

stefkauf committed:
Formatting fixes by pre-commit.

stefkauf commented:
Thanks, @tomaarsen, the problem was a bit mysterious to me. I've fixed the files with pre-commit and updated the pull request.

tomaarsen commented:
I understand! I'm still unsure how to help new contributors with it in an intuitive way.

tomaarsen commented:
Upon reading your commit comment:
> Conversion to Chomsky Normal Form (CNF), implemented as the
'.chomsky_normal_form()' method in grammar.CFG instances, was
incomplete. In particular, it could not deal with empty productions
or mixed productions (with terminals and non-terminals co-occuring
on the right-hand side). The new implementation removes those
shortcomings. Like in the earlier version, the individual steps of
conversion are carried out by calling class methods.

I presume these are related to the following Errors:
https://github.com/nltk/nltk/blob/68e4e589eb62750993aabef5554548973fe1e280/nltk/grammar.py#L746-L749
https://github.com/nltk/nltk/blob/68e4e589eb62750993aabef5554548973fe1e280/nltk/grammar.py#L751-L756

Would it be possible to provide some tests that cover these new cases? 
For example in:
* [nltk/test/grammar.doctest](https://github.com/nltk/nltk/blob/develop/nltk/test/grammar.doctest) (which can be tested with `pytest nltk/test/grammar.doctest`)
* [nltk/test/grammartestsuites.doctest](https://github.com/nltk/nltk/blob/develop/nltk/test/grammartestsuites.doctest) 
* [nltk/test/parse.doctest](https://github.com/nltk/nltk/blob/develop/nltk/test/parse.doctest) 
* or if you prefer pytest, then in a new file under `nltk/test/unit/test_grammar.py`. 

That way, even people like myself who know about Chomsky Normal Form, but don't know the algorithm, can get some confidence in the correctness of the implementation.

stefkauf committed:
Miscellaneous changes to improve readability and documentation.

stefkauf committed:
Changed CFG class to allow for empty list of productions

Context-free grammars can have an empty set of productions.
Conversion to Chomsky Normal Form leads to an empty list of
productions if the language of the CFG is empty (i.e., does not
even contain the empty string - i.e., [S -> S]). The
grammar.CFG class did not allow for this case because some of the
methods to determine properties of the grammar (e.g., self.is_emtpy,
self.is_binary) referred to the minimum or maximum length of the
productions. If the list of productions is emtpy, Python's
'min' and 'max' operators throw an exception. This is fixed now:
the methods determining properties of the grammar are rewritten
using 'all', which succeeds on an empty list. The attributes
self._min_len and self._max_len were left in place and are set to
'None' if there are no productions.

stefkauf commented:
I've been working on some doctest examples. But it is tricky because the requirements for matching outputs are so stringent. For instance, the following grammar for the language a^nb^n:
```    
    S -> 'a' S 'b'
    S -> 
```
was on one run converted to this one, which I put in the doctest file:
```
    S0 -> 
    S0 -> T0 B1
    S0 -> T0 T1
    B0 -> S T1
    B1 -> S T1
    S -> T0 B0
    S -> T0 T1
    T0 -> 'a'
    T1 -> 'b'
```
But sometimes I get a different (but equivalent) result, for instance this one, which differs only in the order in which the new nonterminals were created:
```
    S0 -> 
    S0 -> T1 B0
    S0 -> T1 T0
    B0 -> S T0
    B1 -> S T0
    S -> T1 B1
    S -> T1 T0
    T0 -> 'b'
    T1 -> 'a'
```
There is some arbitrariness in the order in which new nonterminals are created because I use sets. I'm a bit reluctant to change this just to make doctest stop complaining. Would it be acceptable to use some other way to illustrate how it works? For instance, I could put a 'demo' function at the end of grammar.py, for users to invoke.

stefkauf committed:
Secured doctest success

The earlier version created new non-terminals in an unpredictable
order because iteration over sets was used. This led to spurious
failures of doctest. The new version uses order-preserving
containers (dicts or, where possible, lists) to eliminate the
arbitrariness.

stefkauf committed:
Minor formatting changes

iliakur commented:
You may also want to consider unit tests for this. Stick with the ones that give you the biggest bang for the buck: simple to write and can cover more ground. I would argue `pytest` may be a better choice here than doctest.

stefkauf commented:
@iliakur Actually, in my latest update I've solved the problem by using dictionaries instead of sets where the order mattered. Since dictionaries preserve order of insertion, that eliminated the arbitrary variation.

tomaarsen committed:
Merge branch 'develop' into grammar-fixes

tomaarsen commented on nltk/grammar.py:
```diff
@@ -673,8 +674,8 @@ def _calculate_grammar_forms(self):
         self._is_nonlexical = all(p.is_nonlexical() for p in prods if len(p) != 1)
-        self._min_len = min(len(p) for p in prods)
-        self._max_len = max(len(p) for p in prods)
+        self._min_len = min(len(p) for p in prods) if len(prods) else None
```
This doesn't matter much - but both `min` and `max` support `default` keyword arguments: https://docs.python.org/3/library/functions.html#min
These values are used if the provided iterable is empty.
```suggestion
        self._min_len = min((len(p) for p in prods), default=None)
```
(Note, the extra parentheses are because Python'll complain that generator expressions must be parameterized)
(Btw, if you agree with these changes, you can hit "commit changes" on them)

tomaarsen commented on nltk/grammar.py:
```diff
@@ -673,8 +674,8 @@ def _calculate_grammar_forms(self):
-        self._min_len = min(len(p) for p in prods)
-        self._max_len = max(len(p) for p in prods)
+        self._min_len = min(len(p) for p in prods) if len(prods) else None
+        self._max_len = max(len(p) for p in prods) if len(prods) else None
```
Idem.
```suggestion
        self._max_len = max((len(p) for p in prods), default=None)
```

tomaarsen commented:
Well done on adding those doctests! I've handled a merge conflict between this PR and changes introduced from #2888. I'll have a look if I can inspect the implementation itself more thoroughly soon.

tomaarsen reviewed (commented):
As mentioned, I'll make time to look at the rest of the PR too, but this was something I noticed quickly.

stefkauf committed:
Update nltk/grammar.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

stefkauf committed:
Update nltk/grammar.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

tomaarsen commented:
Apologies, I have not yet had time. I've been quite busy.

stevenbird commented on nltk/grammar.py:
```diff
@@ -734,114 +735,456 @@ def is_chomsky_normal_form(self):
 
-    def chomsky_normal_form(self, new_token_padding="@$@", flexible=False):
+    ##################################################
+    # Stefan Kaufmann's proposed changes
```
@stefkauf can we drop such comments please, and just identify you as a co-author at the top of the module?

stevenbird commented on nltk/grammar.py:
```diff
@@ -734,114 +735,456 @@ def is_chomsky_normal_form(self):
+        return cls( grammar.start(), list(new_productions))
+
+
+    # End of Stefan Kaufmann's proposed changes
```
... and this

stevenbird commented on nltk/grammar.py:
```diff
@@ -734,114 +735,456 @@ def is_chomsky_normal_form(self):
+    ##################################################
+
+    ##################################################
+    # Code to be replaced by Stefan Kaufmann's version
```
... and all this

stevenbird commented:
@stefkauf thanks for this great contribution... I've reviewed the code, and just think some of your comments can go, as noted, then we can merge this.

stefkauf committed:
Remove extraneous comments

stefkauf committed:
Merge branch 'grammar-fixes' of https://github.com/stefkauf/nltk into grammar-fixes

Remove extraneous comments

stefkauf commented:
@stevenbird Thanks, Steven, for your messages. I've removed those comments and committed the new version of grammar.py.

tomaarsen committed:
Empty Commit; Force the CI to refresh

rmalouf commented:
What is the status of this @stevenbird?

stefkauf commented:
Hi Rob,

you didn't get a reply about this, did you? I can't tell you what became of it. I have the impression that NLTK is not managed all that actively anymore, but I might be wrong. In any case, I'm not sure.

Hope things are well,
Stefan

---
Department of Linguistics, University of Connecticut
http://stefan-kaufmann.uconn.edu/
________________________________
From: Rob Malouf ***@***.***>
Sent: Tuesday, March 5, 2024 7:12 PM
To: nltk/nltk ***@***.***>
Cc: Kaufmann, Stefan ***@***.***>; Mention ***@***.***>
Subject: Re: [nltk/nltk] Grammar fixes (PR #2884)


*Message sent from a system outside of UConn.*


What is the status of this?

—
Reply to this email directly, view it on GitHub<https://github.com/nltk/nltk/pull/2884#issuecomment-1979848617>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWNJMKBO3GBHCAMYLHMVEX3YWZNQHAVCNFSM5H7O7B5KU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TCOJXHE4DIOBWGE3Q>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

stevenbird commented:
Hi Rob and Stefan I'd welcome an updated PR and will try to turn it around quickly

stefkauf committed:
Merge branch 'develop' into grammar-fixes

stefkauf commented:
I've resolved a merge conflict, it should be good now.

stevenbird commented:
Sorry for the delay!
This is resolved in #3371.
